diff --git a/src/whisper.cpp b/src/whisper.cpp
index 89146e2e..916c92d3 100644
--- a/src/whisper.cpp
+++ b/src/whisper.cpp
@@ -7751,8 +7751,10 @@ int whisper_full_with_state(
                 const int n_segments = state->result_all.size() - n_segments_before;
                 if (ctx->params.dtw_token_timestamps && n_segments) {
                     const int n_frames = std::min(std::min(WHISPER_CHUNK_SIZE * 100, seek_delta), seek_end - seek);
-                    whisper_exp_compute_token_level_timestamps_dtw(
-                            ctx, state, params, result_all.size() - n_segments, n_segments, seek, n_frames, 7, params.n_threads);
+                    if (n_frames >= 4) {
+                        whisper_exp_compute_token_level_timestamps_dtw(
+                                ctx, state, params, result_all.size() - n_segments, n_segments, seek, n_frames, 7, params.n_threads);
+                    }
                     if (params.new_segment_callback) {
                         for (int seg = (int) result_all.size() - n_segments; seg < n_segments; seg++) {
                             params.new_segment_callback(ctx, state, seg, params.new_segment_callback_user_data);
@@ -8969,12 +8971,31 @@ static void median_filter(struct ggml_tensor * dst , const struct ggml_tensor *
     if (ith != 0) {
         return;
     }
+
+    if (a->ne[2] <= 1) {
+        memcpy(dst->data, a->data, ggml_nbytes(a));
+        return;
+    }
+
     int filter_width = ((median_filter_user_data *) userdata)->filter_width;
-    WHISPER_ASSERT(filter_width < a->ne[2]);
     WHISPER_ASSERT(filter_width % 2);
     WHISPER_ASSERT(ggml_n_dims(a) == 3);
     WHISPER_ASSERT(a->type == GGML_TYPE_F32);
 
+    if (filter_width >= a->ne[2]) {
+        filter_width = (int)a->ne[2] - 1;
+        if (filter_width % 2 == 0) {
+            filter_width -= 1;
+        }
+    }
+
+    if (filter_width <= 1) {
+        memcpy(dst->data, a->data, ggml_nbytes(a));
+        return;
+    }
+
+    WHISPER_ASSERT(filter_width < a->ne[2]);
+
     std::vector<float> filter;
     filter.reserve(filter_width);
     for (int64_t i = 0; i < a->ne[0]; ++i) {
@@ -9061,6 +9082,10 @@ static void whisper_exp_compute_token_level_timestamps_dtw(
     WHISPER_ASSERT(state->aheads_cross_QKs != nullptr);
 
     const auto n_audio_tokens = n_frames/2;
+    if (n_audio_tokens <= 1) {
+        ggml_free(gctx);
+        return;
+    }
     WHISPER_ASSERT(state->aheads_cross_QKs != NULL);
     WHISPER_ASSERT(n_audio_tokens <= state->aheads_cross_QKs->ne[1]);
     const auto n_tokens = state->aheads_cross_QKs->ne[0];
